parser/ParseIndexProperties.hpp

```diff
@@ -0,0 +1,267 @@
+/**
+ *   Copyright 2016, Quickstep Research Group, Computer Sciences Department,
+ *     University of Wisconsin—Madison.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ **/
+
+#ifndef QUICKSTEP_PARSER_PARSE_INDEX_PROPERTIES_HPP_
+#define QUICKSTEP_PARSER_PARSE_INDEX_PROPERTIES_HPP_
+
+#include <cstdint>
+#include <memory>
+#include <set>
+#include <string>
+#include <vector>
+#include <unordered_map>
+
+#include "parser/ParseKeyValue.hpp"
+#include "parser/ParseString.hpp"
+#include "parser/ParseTreeNode.hpp"
+#include "storage/StorageBlockLayout.pb.h"
+#include "utility/Macros.hpp"
+#include "utility/PtrList.hpp"
+#include "utility/StringUtil.hpp"
+
+#include "glog/logging.h"
+
+namespace quickstep {
+/** \addtogroup Parser
+ *  @{
+ */
+
+/**
+ * @brief Base class for different types of indices and their associated properties
+ */
+class IndexProperties {
+ public:
+  IndexProperties(bool is_index_description_valid,
+                  const std::string &reason_for_invalid_description,
+                  const ParseKeyValue* invalid_property_node)
+  : is_index_description_valid_(is_index_description_valid),
+    reason_for_invalid_description_(reason_for_invalid_description),
+    invalid_property_node_(invalid_property_node),
+    index_sub_block_description_(nullptr) {
+  }
+
+  typedef std::unordered_map<std::string, ParseKeyValue::KeyValueType> valid_property_map_type;
+
+  bool hasValidIndexDescription() const {
+    return is_index_description_valid_;
+  }
+
+  const ParseKeyValue* getInvalidPropertyNode() const {
+    return invalid_property_node_;
+  }
+
+  std::string getReasonForInvalidIndexDescription() const {
+    return reason_for_invalid_description_;
+  }
+
+  const IndexSubBlockDescription* getIndexDescription() const {
+    return index_sub_block_description_.get();
+  }
+
+  void setIndexDescriptionAsInvalid(const std::string &reason, const ParseKeyValue *invalid_node = nullptr) {
+    is_index_description_valid_ = false;
+    index_sub_block_description_.reset();
+    reason_for_invalid_description_ = reason;
+    invalid_property_node_ = invalid_node;
+  }
+
+  virtual bool addCustomProperties(const PtrList<ParseKeyValue> *key_value_list) = 0;
```

Comment:
@zuyu This function is invoked from the ParseStatement, where a pointer is available to a key_value_list. (returned from the call to custom_properties_node->getKeyValueList()) 
I can de-reference the pointer and pass a reference here. Will it still give better performance?